Update: Replace getTokenBy(Range->Start)Index (refs #1721)

btmills opened the pull request:
The idea for this originated in [a comment on #1721](https://github.com/eslint/eslint/pull/1721#issuecomment-72157629).

The `getTokenByRangeIndex` and the expensive `ranges` array aren't necessary. A space sequence can end one of five ways:

1. It reaches the end of the line.
This rule is not designed to handle end-of-line spaces. Nothing should be reported.

1. It reaches the start of a comment.
The isIndexInComment check will find that pattern.lastIndex is the starting index of the comment. Nothing will be reported.

1. The space sequence is within a comment.
Again, isIndexInComment will prevent reporting.

1. The space sequence is within a string or template literal.
Nothing should be reported.

1. It encounters the start of a token.
In this case, pattern.lastIndex will be exactly the starting index of some token. The rule should report an error.

If I'm not missing anything, then the rule should only report when it encounters a token starting at pattern.lastIndex. Therefore, we replace `getTokenByRangeIndex` with a new method, `getTokenByStartIndex`, consisting simply of `return tokens[starts[index]] || null;`. The rule warns if and only if `getTokenStartingAtIndex(pattern.lastIndex)` returns a token. No more retrieval from the middle of a token's range, and no more need to assemble the map of ranges.

Since this PR builds on #1721, merge that, I'll rebase this so it can be merged, then get #1738 ready and merged?

nzakas commented:
Merged #1721.

btmills commented:
Rebased.

nzakas commented:
I wonder if the name of the method is confusing? Index could refer to the token index (index in the token array), which is why I used rangeIndex. Maybe getTokenByRangeStart?

Also, can you update the Node.js API docs to include this method and remove the old one?

btmills commented:
I like that name much more. Changed.



> Also, can you update the Node.js API docs to include this method and remove the old one?

I didn't see anywhere the old method was documented, but I did find documentation on [the `context` object](http://eslint.org/docs/developer-guide/working-with-rules.html#the-context-object) where they were documented, so I cleaned that up.

nzakas commented:
LGTM
